catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/QueryResultPaginator.java

```diff
@@ -0,0 +1,186 @@
+import java.util.concurrent.LinkedBlockingDeque;
+
+import org.apache.commons.lang3.Validate;
+
+import ddf.catalog.CatalogFramework;
+import ddf.catalog.data.Result;
+import ddf.catalog.federation.FederationException;
+import ddf.catalog.operation.impl.QueryImpl;
+import ddf.catalog.operation.impl.QueryRequestImpl;
+import ddf.catalog.source.SourceUnavailableException;
+import ddf.catalog.source.UnsupportedQueryException;
+
+public class QueryResultPaginator {
+
+    private final CatalogFramework catalogFramework;
+
+    private final QueryImpl queryCopy;
+
+    private int indexOffset;
+
+    private int pageSize;
+
+    //Set to true when a query returns zero results.
+    private boolean isExhausted;
+
+    private BlockingQueue<Result> queriedResultsBuffer;
+
+    /**
+     * QueryResultPaginator takes in the CatalogFramework object and a Query Request and
+     * queries for the number of results needed. This is done using the pageSize specified
+     * in the Query object. The Paginator will return a collection of results of the
+     * specified page size.
+     * <p>
+     * The startIndex is the first index in the results gathered from the CatalogFramework
+     * <p>
+     * The endIndex is the last index in the results gathered from the CatalogFramework, which
+     * may not be reached if the pageSize is smaller than the difference between the endIndex
+     * and startIndex.
+     * <p>
+     * The resultsRemaining are the total results not yet returned in a collection of the size
+     * of a page. This value is tracked through the hasNextPage() method to inform the client that
+     * there are valid results to continue querying.
+     *
+     * @param catalogFramework
+     * @param query
+     */
+    public QueryResultPaginator(CatalogFramework catalogFramework, QueryImpl query) {
+
+        Validate.notNull(catalogFramework);
+        Validate.notNull(query);
+
+        this.catalogFramework = catalogFramework;
+        this.queryCopy = copy(query);
+
+        isExhausted = false;
+        pageSize = queryCopy.getPageSize();
+        indexOffset = queryCopy.getStartIndex();
+
+        Validate.inclusiveBetween(1,
+                Integer.MAX_VALUE,
+                indexOffset,
+                "Query request start index must be greater than zero");
+
+        queriedResultsBuffer = new LinkedBlockingDeque<>();
+    }
+
+    // TODO: 6/27/17 Move this method over to QueryResultIterable inner class
+    /**
+     * If there are pages beyond the current page, returns true.
+     *
+     * @return True if there are results beyond the current page
+     */
+    /**
+     * @return true if there are more results for the client
+     */
+    public boolean hasNext() {
+
+        if (notEmpty()) {
+             return true;
+         }
+
+         if (isExhausted) {
+             return false;
+         }
+
+         //Buffer is empty, but query might not be exhausted.
+         fetchNext();
+         return notEmpty();
```

Comment:
To get here, you had to call `fetchNext()` after having determined previously that the queue had been empty. After your call to `fetchNext()`, one of the following conditions will be true:

1. The queue IS empty AND `isExhausted` is true.
2. The queue IS NOT empty AND `isExhausted` is not true.

Unless you can show me a path by which you've gotten to this line and a different permutation of those two factors has occurred.